accepted/2021/preview-features/preview-features.md

```diff
@@ -0,0 +1,572 @@
+
+This solves two problems:
+
+* The customer is using a future version of C# where a given language feature is
+  no longer considered "preview" and thus doesn't require `LangVersion` to be
+  set to preview but in the targeted runtime the feature is still preview. When
+  a language feature is used that requires the given runtime feature, the
+  compiler should check if the field is marked as preview and fail unless the
+  customer has turned preview mode on.
+
+* The customer has manually set `LangVersion` to `Preview` but
+  `EnablePreviewFeatures` is not configured (thus defaulting to `False`).
+  Similar case as above, the customer can use any language feature that doesn't
+  require runtime changes but as soon as a feature is used that requires a
+  preview runtime feature, the compiler will report an error, demanding that
+  preview mode needs to be turned on.
+
+***Note:*** F# doesn't have the capability to use analyzers today. We can
+decided to make this a compiler feature for F#.
+
+### API Analyzer
+
+We'll need an attribute to record whether or not the project was
+built with preview features turned on:
+
+```C#
+namespace System.Runtime.Versioning
+{
+    [AttributeUsage(AttributeTargets.Assembly |
+                    AttributeTargets.Module |
+                    AttributeTargets.Class |
+                    AttributeTargets.Struct |
+                    AttributeTargets.Enum |
+                    AttributeTargets.Constructor |
+                    AttributeTargets.Method |
+                    AttributeTargets.Property |
+                    AttributeTargets.Field |
+                    AttributeTargets.Event, AllowMultiple = true, Inherited = false)]
+    public sealed class RequiresPreviewFeaturesAttribute : Attribute
+    {
+        public RequiresPreviewFeaturesAttribute();
+    }
+}
+```
+
+We'll ship a built-in analyzer that will report diagnostics when an API is being
+used that is marked `[RequiresPreviewFeatures]` but the consumer (member, type,
+module, assembly) isn't marked. The default severity is `error`.
+
+This analyzer will also validate that if any referenced assembly has applied
+`[RequiresPreviewFeatures]`, that the consuming assembly must also have this
+attribute applied.
+
+User code will typically only have this attribute applied at the assembly-level.
+However, the framework will typically only use this attribute on types and
+members without putting it on the assembly itself.
+
+***Note:*** Since F# doesn't support analyzers, we would have to add this
+capability to the compiler itself.
+
+### Reflection usage
+
+We need to decide whether or not to block calls via reflection to APIs that are
+marked as `[RequirePreviewFeatures]`.
+
+For instance:
+
+```C#
+void f(dynamic x) => x.PreviewMethod();
+```
+
+The problem with doing that is that it's not very pay-for-play friendly (every
+single time a method is called via reflection, we need to check for the applied
```

Comment:
In additional to not being pay-for-play, it is hard to do well. We have tried that multiple times in the past (last time with UWP) and it was always leaky.

Reply:
Good point. I've expanded the section.